R/discrete_normal.R

```diff
@@ -0,0 +1,118 @@
+#' @name discrete_normal
+#' @title Discrete normal distribution
+#'
+#' @description a discretised normal distribution (i.e. sampled by applying 
+#'   the round operation to samples from a normal). Due to the numerical 
+#'   instability of integrating across the distribution, a vector of breaks 
+#'   must be defined and the observations will be treated as censored 
+#'   within those breaks
+#'
+#' @param mean unconstrained parameters giving the mean of the distribution
+#' @param sd unconstrained parameters giving the standard deviation of the 
+#'   distribution
+#' @param breaks a vector of breaks; observations will be treated as censored 
+#'   within those breaks
+#' @param dim a scalar giving the number of rows in the resulting greta array
+#'
+#' @importFrom R6 R6Class
+#' @export
+
+discrete_normal <- function(mean, sd, breaks, dim = NULL) {
+  distrib("discrete_normal", mean, sd, breaks, dim)
+}
+
+# define the discrete normal distribution
+discrete_normal_distribution <- R6Class(
+  classname = "discrete_normal_distribution",
+  inherit = distribution_node,
+  public = list(
+    
+    breaks = NA,
+    lower_bounds = NA,
+    upper_bounds = NA,
+    
+    initialize = function(mean, sd, breaks, dim) {
+      
+      mean <- as.greta_array(mean)
+      sd   <- as.greta_array(sd)
+      
+      # check length of breaks
+      if (length(breaks) <= 1) {
+        msg <- cli::format_error(
+          c(
+            "{.var breaks} must be a vector with at least two break points",
+            "but {.var breaks} has length {length(breaks)}"
+          )
+        ) 
+        stop(
+          msg,
+          call. = FALSE
+        )
+      }
+      
+      # handle gradient issue between sdlog and 0s
```

Comment:
you may not need this at all - I think this was only in the lognormal distribution because of a discrepancy between how TF and R evaluate the lognormal density at 0 (or lower). In R `dlnorm(0)` evaluates to 0, in TF I think it's either an error or NaN. That shouldn't be a problem for normal, since it has support on all real numbers, not just positive (or non-zero) numbers.

Reply:
oh hang on. Now that I think about it, leaving this in could be what is biasing your estimates of the parameters, since the model gains no knowledge of any values lower than ~0. So maybe deleting the next line will help

Reply:
Thanks for taking a look. If I simply delete https://github.com/greta-dev/greta.distributions/blob/e2ae5b6b45a05027cfeed05dd9ab480791cb6f02/R/discrete_normal.R#L53-L54 it'll error 
```
Error in py_call_impl(callable, dots$args, dots$keywords) : 
  tensorflow.python.framework.errors_impl.InvalidArgumentError: indices[0,2739,0] = -2147483648 is not in [0, 50)
<... omitted ...>rom an input operation.
Input Source operations connected to node GatherV2:
 Const_4 (defined at /framework/constant_op.py:180)

Original stack trace for 'GatherV2':
  File "/util/dispatch.py", line 180, in wrapper
    return target(*args, **kwargs)
  File "/ops/array_ops.py", line 3475, in gather
    return gen_array_ops.gather_v2(params, indices, axis, name=name)
  File "/ops/gen_array_ops.py", line 4097, in gather_v2
    batch_dims=batch_dims, name=name)
  File "/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/framework/ops.py", line 3616, in create_op
    op_def=op_def)
  File "/framework/ops.py", line 2005, in __init__
    self._traceback = tf_stack.extract_stack()

See `reticulate::py_last_error()` for details
```
From the top lines it seems that we have negative values when we shouldn't have.

Sorry I am not familiar with the tensorflow codes but will try my best to fix. 

I'm also not sure if the CDF section could be causing problem, esp. this:
https://github.com/greta-dev/greta.distributions/blob/e2ae5b6b45a05027cfeed05dd9ab480791cb6f02/R/discrete_normal.R#L87

For completeness, with the latest commit, the code below will get a posterior of `mulog` and `sdlog` quite far from the default 0 and 1 in `rlnorm`, respectively. (also unsure if the unintended constrain also make the sampler got stuck sometimes) From the code:

```
library(greta.distributions)

y   <- rlnorm(10000)
y_r <- round(y)
mulog <- normal(0, 10)
sdlog <- exponential(10)
distribution(y_r) <- discrete_lognormal(mulog, sdlog, breaks = seq(0, ceiling(max(y)), 1))  # not sure if max(y) is legit

mod <- model(mulog, sdlog)

draws <- mcmc(mod, warmup = 2000)
plot(draws)
```
the posteriors look like this

![image](https://user-images.githubusercontent.com/20037698/182492295-31b3c3f9-4b66-44cb-9374-7451d78b868d.png)

Reply:
I'm confused. Did you make these changes to the lognormal instead of the normal?

Reply:
sorry, yes I made these changes to lognormal, not normal. I copied the snippets from the wrong script. Just to be clear, I deleted `breaks <- pmax(breaks, .Machine$double.eps)` in the lognormal case.

Reply:
OK, that makes sense. The `pmax(breaks, ...)` code needs to be there in the lognormal case to handle evaluation at 0, but I think not in the normal case.

Reply:
alright, thanks for clarifying again. I removed `pmax(breaks, ...)` from the normal case, then I ran the following test for normal:
```
library(greta.distributions)

y <- rnorm(10000)
y_r <- round(y)
mu <- normal(0, 10)
sd <- exponential(1)
distribution(y_r) <- discrete_normal(mu, sd, breaks = seq(floor(min(y)), ceiling(max(y)), 1))

mod <- model(mu, sd)

draws <- mcmc(mod)
plot(draws)
```
but the posterior mean and sd are far from 0 and 1, respectively.

![image](https://user-images.githubusercontent.com/20037698/182506322-72dc1bed-9d2b-4f8e-9d43-bf9f260b94e2.png)

Could this be due to the `x_safe <- tf$math$maximum(x, fl(.Machine$double.eps))` part?

Also, if instead I shift the true normal mean to more negative and simulate data as `y <- rnorm(10000, -10)`, I also got a similar error as in the lognormal case:
```
Error in py_call_impl(callable, dots$args, dots$keywords) : 
  tensorflow.python.framework.errors_impl.InvalidArgumentError: indices[0,0,0] = -2147483648 is not in [0, 8)
	 [[node GatherV2 (defined at /util/dispatch.py:180) ]]

Errors may have originated from an input operation.
Input Source operations connected to node GatherV2:
 Const_4 (defined at /framework/constant_op.py:180)

Original stack trace for 'GatherV2':
  File "/util/dispatch.py", line 180, in wrapper
    return target(*args, **kwargs)
  File "/ops/array_ops.py", line 3475, in gather
    return gen_array_ops.gather_v2(params, indices, axis, name=name)
  File "/ops/gen_array_ops.py", line 4097, in gather_v2
    batch_dims=batch_dims, name=name)
  File "/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/framework/ops.py", line 3616, in create_op
    op_def=op_def)
  File "/framework/ops.py", line 2005, in __init__
    self._traceback = tf_stack.extract_s
```

Reply:
Ah yes, good catch. I see now that you had flagged both of these already! tf_safe_cdf is hard-coded to have (0,Inf) support. It will need to be modified to pass in the bounds of the distribution. So the fl(0) bit will need to be changed to fl(lower_bound), and the lower bound passed in via an argument. For completeness, the upper bound should be passed in too.

Reply:
Tried to implement your suggestion as well as removing `x_safe <- tf$math$maximum(x, fl(.Machine$double.eps))` in [15cc772](https://github.com/greta-dev/greta.distributions/pull/16/commits/15cc772c07238b4ce5e3345de8a1c2a0dbd7a364) , though not 100% sure what I'm doing 😅 

Now sampling the model below
```
y <- rnorm(10000, 0, 1)
y_r <- round(y)
mu <- normal(0, 10)
sd <- exponential(1)
distribution(y_r) <- discrete_normal(mu, sd, breaks = seq(floor(min(y_r))-1, ceiling(max(y_r))+1, 1))
mod <- model(mu, sd)
```
yields the posteriors
![image](https://user-images.githubusercontent.com/20037698/182515599-bf64736c-17ef-4bfc-aa32-792a059307dd.png)
Now `sd` is so much closer to 1 (though mildly positively biased in all examples below), but `mu` has a curious mode near 0.5.

It gets curious-er if we shift the true mean to, say, 2 with `y <- rnorm(10000, 2, 1)`:
![image](https://user-images.githubusercontent.com/20037698/182516228-997ff30e-4d94-45cd-a739-cf97ce892edc.png)

If we shift the true mean to more negative `y <- rnorm(10000, -10, 1)`:
![image](https://user-images.githubusercontent.com/20037698/182516551-c614dc58-d8f3-4bfd-b245-c4c63273ba8d.png)
It seems that the estimated `mu` tends to be positively bias for 0.5... 

Maybe I made some mistakes in `tf_safe_cdf`...

p/s: not sure if this is related, but sometimes large mean or sd magnitude (e.g., `y <- rnorm(10000, 0, 10)` or `y <- rnorm(10000, 10, 1)`) will result in 
```
> draws <- mcmc(mod)
Error: Could not find reasonable starting values after 20 attempts.
Please specify initial values manually via the `initial_values` argument
```
but from the error message I wasn't able to debug further...

Reply:
sorry I just want to quickly add if this might be helpful: the 0.5 bias could be due to the way I specifiy the integer breaks `seq(from, to, by = 1)`. Given that an unrounded variable `z` is related to a rounded variable `y` as `z \in (y-0.5, y+0.5)`... not sure if this is relevant.

Reply:
I think the culprit may be `breaks` and how it is used in `tf_safe_cdf`. Consider a continuous variable that is recorded/rounded as integers `c(-2, -1, 0, 1, 2)`. When we specify `discrete_normal(..., breaks = c(-2, -1, 0, 1, 2))`, `tf_safe_cdf` will evaluate the CDF at these breaks, but don't we want it to evaluate at `breaks - 0.5` and `breaks + 0.5`? In other words, evaluating the CDF at `breaks` implies that the discretisation happens *between* rather than *at* the breaks?

I pushed an experimental `discrete_normal` that also requires an additional argument `edges`, which have one more element than `breaks` and are the bounds of each `breaks` values. For the example above we would specify `edges = c(-2.5, -1.5 -0.5, 0.5, 1.5, 2.5)`. This feels too hard-coded / manual / hard to explain at the moment, and my code can probably be refactored. But playing with it yields a posterior of `mean` that is much closer to the empirical mean:
```
library(greta.distributions)

y <- rnorm(10000, 0, 1)
y_r <- round(y)
mu <- normal(0, 10)
sd <- exponential(1)
breaks_diff <- 1
breaks <- seq(floor(min(y_r)), ceiling(max(y_r)), breaks_diff)
edges_min <- min(y_r) - breaks_diff/2
edges_max <- breaks[length(breaks)] + breaks_diff/2
edges <- c(edges_min, breaks[-length(breaks)] + breaks_diff/2, edges_max)
distribution(y_r) <- discrete_normal(mu, sd, breaks = breaks, edges = edges)
mod <- model(mu, sd)
draws <- mcmc(mod, warmup = 2000)
plot(draws)
```
![image](https://user-images.githubusercontent.com/20037698/182741622-7b2b43ef-e315-4406-b491-c95f7068ac44.png)

Compare the empirical mean to the posterior summary.
```
> mean(y)
[1] -0.006322464

> summary(draws)

Iterations = 1:1000
Thinning interval = 1 
Number of chains = 4 
Sample size per chain = 1000 

1. Empirical mean and standard deviation for each variable,
   plus standard error of the mean:

        Mean       SD  Naive SE Time-series SE
mu -0.008201 0.010377 0.0001641      0.0001913
sd  1.002026 0.007522 0.0001189      0.0001978

2. Quantiles for each variable:

       2.5%      25%       50%       75%   97.5%
mu -0.02856 -0.01492 -0.008117 -0.001294 0.01179
sd  0.98783  0.99688  1.002065  1.007069 1.01686
```